File: webservice.java
import com.google.gson.Gson; 
import com.sun.org.apache.xml.internal.security.utils.Base64; 
 
import javax.crypto.Mac; 
import javax.crypto.spec.SecretKeySpec; 
import java.io.BufferedReader; 
import java.io.DataOutputStream; 
import java.io.IOException; 
import java.io.InputStreamReader; 
import java.net.HttpURLConnection; 
import java.net.URL; 
import java.security.InvalidKeyException; 
import java.security.MessageDigest; 
import java.security.NoSuchAlgorithmException; 
import java.text.SimpleDateFormat; 
import java.util.Date; 
import java.util.HashMap; 
import java.util.TimeZone; 

class Main { 
 
    public static HashMap<String, String> getAuthorizationHeaders(String method,String path ,String body, String merchantKey, String merchantSalt) throws IOException, NoSuchAlgorithmException, InvalidKeyException { 
        try { 
            HashMap<String, String> headers = new HashMap<String, String>(); 
            Date today = new Date(); 
            SimpleDateFormat sdf = new SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss zzz"); 
            sdf.setTimeZone(TimeZone.getTimeZone("GMT")); 
            String date = sdf.format(today); 
            byte[] digest = MessageDigest.getInstance("SHA-256").digest(body.getBytes()); 
            String digestData = Base64.encode(digest); 
            String signingString = "date: " + date + "\n" + "digest: " + digestData; 
            SecretKeySpec signingKey = new SecretKeySpec(merchantSalt.getBytes(), "HmacSHA256"); 
            Mac mac = Mac.getInstance("HmacSHA256"); 
            mac.init(signingKey); 
            byte[] result = mac.doFinal(signingString.getBytes("ASCII")); 
            String hash = Base64.encode(result); 
            String authorization = "hmac username=" + "\"" + merchantKey + "\", algorithm=" + "\"hmac-sha256\", headers=" + "\"" + "date digest" + "\", signature=" + "\"" + hash + "\""; 
            headers.put("Date", date); 
            headers.put("Digest", digestData); 
            headers.put("Authorization", authorization); 
            System.out.println("header :"+headers); 
            return headers; 
        } catch(Exception e) { 
            e.printStackTrace(); 
        } 
        return null; 
    } 
 

    public static void sendWebService(String data,String path,String method) throws Exception { 
        String url = "https://sandbox.payu.in"+path; 
        URL obj = new URL(url); 
        HttpURLConnection con = (HttpURLConnection) obj.openConnection(); 
        HashMap headers=getAuthorizationHeaders(method,path,data,"smsplus","ddwefw"); 
        con.setRequestMethod(method); 
        con.setRequestProperty("Content-Type", "application/json"); 
        con.setRequestProperty("authorization", String.valueOf(headers.get("Authorization"))); 
        con.setRequestProperty("date", String.valueOf(headers.get("Date"))); 
        con.setRequestProperty("digest", String.valueOf(headers.get("Digest"))); 
        //con.setRequestProperty("host", "api.payu.in"); 
 
        if(method.equalsIgnoreCase("POST") && !data.isEmpty() && data!=null){ 
            con.setDoOutput(true); 
            byte[] outputBytes = data.getBytes("UTF-8");  
            OutputStream os = con.getOutputStream();  
            os.write(outputBytes);  
            os.flush(); 
            os.close(); 
        } 
 
        int responseCode = con.getResponseCode(); 
        System.out.println("\nSending request to URL : " + url); 
        System.out.println("Parameters : " + data); 
        System.out.println("Response Code : " + responseCode); 
 
 
        BufferedReader in; 
        if(responseCode!=200){ 
            in = new BufferedReader(new InputStreamReader(con.getErrorStream())); 
        }else{ 
            in = new BufferedReader(new InputStreamReader(con.getInputStream())); 
        } 
 
        String inputLine; 
        StringBuffer response = new StringBuffer(); 
 
        while ((inputLine = in.readLine()) != null) { 
            response.append(inputLine); 
        } 
        in.close(); 
 
        System.out.println(response.toString()); 
 
    }

}

File: webservice.php
public static function getHmacParameters($method, $path, $content, $username, $salt) { 
 
    $dateString = gmdate('D, d M Y H:i:s T'); 
    $digest = base64_encode(hash("sha256", $content, true)); 
 
    $signing_headers = [ 
        'date' => $dateString, 
        'digest' => $digest 
    ]; 
 
    $signing_string = ""; 
    $headers = ""; 
    foreach ($signing_headers as $key => $value) { 
        $signing_string .= $key . ":" . " "; 
        $signing_string .= $value . "\n"; 
        $headers .= $key . " "; 
    } 
    $signing_string = rtrim($signing_string, "\n"); 
    $headers = rtrim($headers, " "); 
 
    $hmacHash = hash_hmac("sha256", $signing_string, $salt, true); 
 
    $signature = base64_encode($hmacHash); 
    $authorization = "hmac username=\"$username\", algorithm=\"hmac-sha256\", headers=\"$headers\", signature=\"$signature\""; 
 
    return [ 
        'dateString' => $dateString, 
        'digest' => $digest, 
        'authorization' => $authorization 
    ]; 
 
} 


public static function sendWebService($merchantKey,$path, $reqBody,$method, $merchantSalt) { 
         
if($reqBody==null ){ 
    $reqBody = ''; 
} 
 
$url = 'https://sandbox.payu.in' . $path; 
 
getHmacParameters($method, $path, $reqBody, $merchantKey, $merchantSalt); 
 
$ch = curl_init() or die(curl_error($ch)); 
curl_setopt($ch, CURLOPT_URL, $url); 
 
$headers = array( 
    "date: " . $hmacParams['dateString'], 
    "digest: " . $hmacParams['digest'], 
    "authorization: " . $hmacParams['authorization'] 
); 
 
if ($method == 'POST') { 
    curl_setopt($ch, CURLOPT_POST, 1); 
} else if ($method != 'GET'){ 
    curl_setopt($ch, CURLOPT_CUSTOMREQUEST, $method); 
} 
 
if ($reqBody!='' && $method != 'GET') { 
    curl_setopt($ch, CURLOPT_POSTFIELDS, $reqBody); 
    $headers[] = 'Content-Type: application/json'; 
} 
curl_setopt($ch, CURLOPT_HTTPHEADER, $headers); 
 
 
curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1); 
curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, 0); 
curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, 0); 
curl_setopt($ch, CURLOPT_HEADER, 1); 
//        curl_setopt($ch, CURLOPT_TIMEOUT, 10); 
 
$response = curl_exec($ch); 
list($responseHeader, $content) = explode("\r\n\r\n", $response, 2); 
$responseHeaders = explode("\n", $responseHeader); 
 
$info = curl_getinfo($ch); 
 
$statusLabel = "Http status code: " . $info["http_code"]; 
 
$curlErrorNo = curl_errno($ch); 
if ($curlErrorNo) { 
    $message = curl_error($ch); 
    $statusLabel = "Curl call failed!. Error-code: " . $curlErrorNo . " Error-msg: " . $message; 
} 
 
echo $content; 
         
} 


